File: gr/network-visualization/network_layout/layout/labelAdjust/LabelAdjust.cs
using System;
using System.Collections.Generic;

/*
 Copyright 2008-2010 Gephi
 Authors : Mathieu Jacomy
 Website : http://www.gephi.org

 This file is part of Gephi.

 DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.

 Copyright 2011 Gephi Consortium. All rights reserved.

 The contents of this file are subject to the terms of either the GNU
 General Public License Version 3 only ("GPL") or the Common
 Development and Distribution License("CDDL") (collectively, the
 "License"). You may not use this file except in compliance with the
 License. You can obtain a copy of the License at
 http://gephi.org/about/legal/license-notice/
 or /cddl-1.0.txt and /gpl-3.0.txt. See the License for the
 specific language governing permissions and limitations under the
 License.  When distributing the software, include this License Header
 Notice in each file and include the License files at
 /cddl-1.0.txt and /gpl-3.0.txt. If applicable, add the following below the
 License Header, with the fields enclosed by brackets [] replaced by
 your own identifying information:
 "Portions Copyrighted [year] [name of copyright owner]"

 If you wish your version of this file to be governed by only the CDDL
 or only the GPL Version 3, indicate your decision by adding
 "[Contributor] elects to include this software in this distribution
 under the [CDDL or GPL Version 3] license." If you do not indicate a
 single choice of license, a recipient has the option to distribute
 your version of this file under either the CDDL, the GPL Version 3 or
 to extend the choice of license to its licensees as provided above.
 However, if you add GPL Version 3 code and therefore, elected the GPL
 Version 3 license, then the option applies only if the new code is
 made subject to such option by the copyright holder.

 Contributor(s):

 Portions Copyrighted 2011 Gephi Consortium.
 */

namespace org.gephi.layout.plugin.labelAdjust
{
	using Graph = org.gephi.graph.api.Graph;
	using Node = org.gephi.graph.api.Node;
	using TextProperties = org.gephi.graph.api.TextProperties;
	using Layout = org.gephi.layout.spi.Layout;
	using LayoutBuilder = org.gephi.layout.spi.LayoutBuilder;
	using LayoutProperty = org.gephi.layout.spi.LayoutProperty;
	using Exceptions = org.openide.util.Exceptions;
	using NbBundle = org.openide.util.NbBundle;

	/// <summary>
	/// @author Mathieu Jacomy
	/// </summary>
	public class LabelAdjust : AbstractLayout, Layout
	{

		//Graph
		protected internal Graph graph;
		//Settings
		private double speed = 1;
		private bool adjustBySize = true;
		private float radiusScale = 1.1f;
		//Graph size
		private float xmin;
		private float xmax;
		private float ymin;
		private float ymax;

		public LabelAdjust(LayoutBuilder layoutBuilder) : base(layoutBuilder)
		{
		}

		public override void resetPropertiesValues()
		{
			speed = 1;
			radiusScale = 1.1f;
			adjustBySize = true;
		}

		public override void initAlgo()
		{
			Converged = false;
		}

		public override void goAlgo()
		{
			this.graph = graphModel.GraphVisible;
			graph.readLock();
			try
			{
				Node[] nodes = graph.Nodes.toArray();

				//Reset Layout Data
				foreach (Node n in nodes)
				{
					if (n.LayoutData == null || !(n.LayoutData is LabelAdjustLayoutData))
					{
						n.LayoutData = new LabelAdjustLayoutData();
					}
					LabelAdjustLayoutData layoutData = n.LayoutData;
					layoutData.freeze = 0;
					layoutData.dx = 0;
					layoutData.dy = 0;
				}

				// Get xmin, xmax, ymin, ymax
				xmin = float.MaxValue;
				xmax = float.Epsilon;
				ymin = float.MaxValue;
				ymax = float.Epsilon;

				IList<Node> correctNodes = new List<Node>();
				foreach (Node n in nodes)
				{
					float x = n.x();
					float y = n.y();
					TextProperties t = n.TextProperties;
					float w = t.Width;
					float h = t.Height;
					float radius = n.size() / 2f;

					if (w > 0 && h > 0)
					{
						// Get the rectangle occupied by the node (size + label)
						float nxmin = Math.Min(x - w / 2, x - radius);
						float nxmax = Math.Max(x + w / 2, x + radius);
						float nymin = Math.Min(y - h / 2, y - radius);
						float nymax = Math.Max(y + h / 2, y + radius);

						// Update global boundaries
						xmin = Math.Min(this.xmin, nxmin);
						xmax = Math.Max(this.xmax, nxmax);
						ymin = Math.Min(this.ymin, nymin);
						ymax = Math.Max(this.ymax, nymax);

						correctNodes.Add(n);
					}
				}

				if (correctNodes.Count == 0 || xmin == xmax || ymin == ymax)
				{
					return;
				}

				long timeStamp = 1;
				bool someCollision = false;

				//Add all nodes in the quadtree
				QuadTree quadTree = new QuadTree(this, correctNodes.Count, (xmax - xmin) / (ymax - ymin));
				foreach (Node n in correctNodes)
				{
					quadTree.add(n);
				}

				//Compute repulsion - with neighbours in the 8 quadnodes around the node
				foreach (Node n in correctNodes)
				{
					timeStamp++;
					LabelAdjustLayoutData layoutData = n.LayoutData;
					QuadNode quad = quadTree.getQuadNode(layoutData.labelAdjustQuadNode);

					//Repulse with adjacent quad - but only one per pair of nodes, timestamp is guaranteeing that
					foreach (Node neighbour in quadTree.getAdjacentNodes(quad.row, quad.col))
					{
						LabelAdjustLayoutData neighborLayoutData = neighbour.LayoutData;
						if (neighbour != n && neighborLayoutData.freeze < timeStamp)
						{
							bool collision = repulse(n, neighbour);
							someCollision = someCollision || collision;
						}
						neighborLayoutData.freeze = timeStamp; //Use the existing freeze float variable to set timestamp
					}
				}

				if (!someCollision)
				{
					Converged = true;
				}
				else
				{
					// apply forces
					foreach (Node n in correctNodes)
					{
						LabelAdjustLayoutData layoutData = n.LayoutData;
						if (!n.Fixed)
						{
							layoutData.dx *= (float)speed;
							layoutData.dy *= (float)speed;
							float x = n.x() + layoutData.dx;
							float y = n.y() + layoutData.dy;

							n.X = x;
							n.Y = y;
						}
					}
				}
			}
			finally
			{
				graph.readUnlockAll();
			}
		}

		private bool repulse(Node n1, Node n2)
		{
			bool collision = false;
			float n1x = n1.x();
			float n1y = n1.y();
			float n2x = n2.x();
			float n2y = n2.y();
			TextProperties t1 = n1.TextProperties;
			TextProperties t2 = n2.TextProperties;
			float n1w = t1.Width;
			float n2w = t2.Width;
			float n1h = t1.Height;
			float n2h = t2.Height;
			LabelAdjustLayoutData n2Data = n2.LayoutData;

			double n1xmin = n1x - 0.5 * n1w;
			double n2xmin = n2x - 0.5 * n2w;
			double n1ymin = n1y - 0.5 * n1h;
			double n2ymin = n2y - 0.5 * n2h;
			double n1xmax = n1x + 0.5 * n1w;
			double n2xmax = n2x + 0.5 * n2w;
			double n1ymax = n1y + 0.5 * n1h;
			double n2ymax = n2y + 0.5 * n2h;

			//Sphere repulsion
			if (adjustBySize)
			{
				double xDist = n2x - n1x;
				double yDist = n2y - n1y;
				double dist = Math.Sqrt(xDist * xDist + yDist * yDist);
				bool sphereCollision = dist < radiusScale * (n1.size() + n2.size());
				if (sphereCollision)
				{
					double f = 0.1 * n1.size() / dist;
					if (dist > 0)
					{
						n2Data.dx = (float)(n2Data.dx + xDist / dist * f);
						n2Data.dy = (float)(n2Data.dy + yDist / dist * f);
					}
					else
					{
						n2Data.dx = (float)(n2Data.dx + 0.01 * (0.5 - GlobalRandom.NextDouble));
						n2Data.dy = (float)(n2Data.dy + 0.01 * (0.5 - GlobalRandom.NextDouble));
					}
					collision = true;
				}
			}

			double upDifferential = n1ymax - n2ymin;
			double downDifferential = n2ymax - n1ymin;
			double labelCollisionXleft = n2xmax - n1xmin;
			double labelCollisionXright = n1xmax - n2xmin;

			if (upDifferential > 0 && downDifferential > 0)
			{ // Potential collision
				if (labelCollisionXleft > 0 && labelCollisionXright > 0)
				{ // Collision
					if (upDifferential > downDifferential)
					{
						// N1 pushes N2 up
						n2Data.dy = (float)(n2Data.dy - 0.02 * n1h * (0.8 + 0.4 * GlobalRandom.NextDouble));
						collision = true;
					}
					else
					{
						// N1 pushes N2 down
						n2Data.dy = (float)(n2Data.dy + 0.02 * n1h * (0.8 + 0.4 * GlobalRandom.NextDouble));
						collision = true;
					}
					if (labelCollisionXleft > labelCollisionXright)
					{
						// N1 pushes N2 right
						n2Data.dx = (float)(n2Data.dx + 0.01 * (n1h * 2) * (0.8 + 0.4 * GlobalRandom.NextDouble));
						collision = true;
					}
					else
					{
						// N1 pushes N2 left
						n2Data.dx = (float)(n2Data.dx - 0.01 * (n1h * 2) * (0.8 + 0.4 * GlobalRandom.NextDouble));
						collision = true;
					}
				}
			}

			return collision;
		}

		public override void endAlgo()
		{
			foreach (Node n in graph.Nodes)
			{
				n.LayoutData = null;
			}
		}

		public override LayoutProperty[] Properties
		{
			get
			{
				IList<LayoutProperty> properties = new List<LayoutProperty>();
				const string LABELADJUST_CATEGORY = "LabelAdjust";
				try
				{
					properties.Add(LayoutProperty.createProperty(this, typeof(Double), NbBundle.getMessage(this.GetType(), "LabelAdjust.speed.name"), LABELADJUST_CATEGORY, "LabelAdjust.speed.name", NbBundle.getMessage(this.GetType(), "LabelAdjust.speed.desc"), "getSpeed", "setSpeed"));
					properties.Add(LayoutProperty.createProperty(this, typeof(Boolean), NbBundle.getMessage(this.GetType(), "LabelAdjust.adjustBySize.name"), LABELADJUST_CATEGORY, "LabelAdjust.adjustBySize.name", NbBundle.getMessage(this.GetType(), "LabelAdjust.adjustBySize.desc"), "isAdjustBySize", "setAdjustBySize"));
				}
				catch (Exception e)
				{
					Exceptions.printStackTrace(e);
				}
				return ((List<LayoutProperty>)properties).ToArray();
			}
		}

		public virtual double? Speed
		{
			get
			{
				return speed;
			}
			set
			{
				this.speed = value.Value;
			}
		}


		public virtual bool? AdjustBySize
		{
			get
			{
				return adjustBySize;
			}
			set
			{
				this.adjustBySize = value.Value;
			}
		}


		private class QuadNode
		{

			internal readonly int index;
			internal readonly int row;
			internal readonly int col;
			internal readonly IList<Node> nodes;

			public QuadNode(int index, int row, int col)
			{
				this.index = index;
				this.row = row;
				this.col = col;
				this.nodes = new List<Node>();
			}

			public virtual IList<Node> Nodes
			{
				get
				{
					return nodes;
				}
			}

			public virtual void add(Node n)
			{
				nodes.Add(n);
			}
		}

		private class QuadTree
		{
			private readonly LabelAdjust outerInstance;


			internal readonly QuadNode[] quads;
			internal readonly int COLUMNS;
			internal readonly int ROWS;

			public QuadTree(LabelAdjust outerInstance, int numberNodes, float aspectRatio)
			{
				this.outerInstance = outerInstance;
				if (aspectRatio > 0)
				{
					COLUMNS = (int) Math.Ceiling(numberNodes / 50f);
					ROWS = (int) Math.Ceiling(COLUMNS / aspectRatio);
				}
				else
				{
					ROWS = (int) Math.Ceiling(numberNodes / 50f);
					COLUMNS = (int) Math.Ceiling(ROWS / aspectRatio);
				}
				quads = new QuadNode[COLUMNS * ROWS];
				for (int row = 0; row < ROWS; row++)
				{
					for (int col = 0; col < COLUMNS; col++)
					{
						quads[row * COLUMNS + col] = new QuadNode(row * COLUMNS + col, row, col);
					}
				}
			}

			public virtual void add(Node node)
			{
				float x = node.x();
				float y = node.y();
				TextProperties t = node.TextProperties;
				float w = t.Width;
				float h = t.Height;
				float radius = node.size();

				// Get the rectangle occupied by the node (size + label)
				float nxmin = Math.Min(x - w / 2, x - radius);
				float nxmax = Math.Max(x + w / 2, x + radius);
				float nymin = Math.Min(y - h / 2, y - radius);
				float nymax = Math.Max(y + h / 2, y + radius);

				// Get the rectangle as boxes
				int minXbox = (int) Math.Floor((COLUMNS - 1) * (nxmin - outerInstance.xmin) / (outerInstance.xmax - outerInstance.xmin));
				int maxXbox = (int) Math.Floor((COLUMNS - 1) * (nxmax - outerInstance.xmin) / (outerInstance.xmax - outerInstance.xmin));
				int minYbox = (int) Math.Floor((ROWS - 1) * (((outerInstance.ymax - outerInstance.ymin) - (nymax - outerInstance.ymin)) / (outerInstance.ymax - outerInstance.ymin)));
				int maxYbox = (int) Math.Floor((ROWS - 1) * (((outerInstance.ymax - outerInstance.ymin) - (nymin - outerInstance.ymin)) / (outerInstance.ymax - outerInstance.ymin)));
				for (int col = minXbox; col <= maxXbox && col < COLUMNS && col >= 0; col++)
				{
					for (int row = minYbox; row <= maxYbox && row < ROWS && row >= 0; row++)
					{
						quads[row * COLUMNS + col].add(node);
					}
				}

				//Get the node center
				int centerX = (int) Math.Floor((COLUMNS - 1) * (x - outerInstance.xmin) / (outerInstance.xmax - outerInstance.xmin));
				int centerY = (int) Math.Floor((ROWS - 1) * (((outerInstance.ymax - outerInstance.ymin) - (y - outerInstance.ymin)) / (outerInstance.ymax - outerInstance.ymin)));
				LabelAdjustLayoutData layoutData = node.LayoutData;
				layoutData.labelAdjustQuadNode = quads[centerY * COLUMNS + centerX].index;
			}

			public virtual IList<Node> get(int row, int col)
			{
				return quads[row * ROWS + col].Nodes;
			}

			public virtual IList<Node> getAdjacentNodes(int row, int col)
			{
				if (quads.Length == 1)
				{
					return quads[0].Nodes;
				}

				IList<Node> adjNodes = new List<Node>();
				int left = Math.Max(0, col - 1);
				int top = Math.Max(0, row - 1);
				int right = Math.Min(COLUMNS - 1, col + 1);
				int bottom = Math.Min(ROWS - 1, row + 1);
				for (int i = left; i <= right; i++)
				{
					for (int j = top; j <= bottom; j++)
					{
						((List<Node>)adjNodes).AddRange(quads[j * COLUMNS + i].Nodes);
					}
				}
				return adjNodes;
			}

			public virtual QuadNode getQuadNode(int index)
			{
				return quads[index];
			}
		}
	}

}

File: gr/physics/layout/LabelAdjust.vb
#Region "Microsoft.VisualBasic::22d529a3b06becec315d4644ed6f410a, gr\physics\layout\LabelAdjust.vb"

    ' Author:
    ' 
    '       asuka (amethyst.asuka@gcmodeller.org)
    '       xie (genetics@smrucc.org)
    '       xieguigang (xie.guigang@live.com)
    ' 
    ' Copyright (c) 2018 GPL3 Licensed
    ' 
    ' 
    ' GNU GENERAL PUBLIC LICENSE (GPL3)
    ' 
    ' 
    ' This program is free software: you can redistribute it and/or modify
    ' it under the terms of the GNU General Public License as published by
    ' the Free Software Foundation, either version 3 of the License, or
    ' (at your option) any later version.
    ' 
    ' This program is distributed in the hope that it will be useful,
    ' but WITHOUT ANY WARRANTY; without even the implied warranty of
    ' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    ' GNU General Public License for more details.
    ' 
    ' You should have received a copy of the GNU General Public License
    ' along with this program.  If not, see <http://www.gnu.org/licenses/>.
    
    
    
    ' /********************************************************************************/

    ' Code Statistics:

    '   Total Lines: 424
    '    Code Lines: 284 (66.98%)
    ' Comment Lines: 79 (18.63%)
    '    - Xml Docs: 25.32%
    ' 
    '   Blank Lines: 61 (14.39%)
    '     File Size: 17.16 KB


    '     Class LabelAdjust
    ' 
    '         Properties: AdjustBySize, Speed
    ' 
    '         Function: Repulse
    ' 
    '         Sub: Iterate, ResetPropertiesValues, Solve
    ' 
    ' /********************************************************************************/

#End Region

Imports System.Drawing
Imports randf = Microsoft.VisualBasic.Math.RandomExtensions
Imports std = System.Math

' 
' Copyright 2008-2010 Gephi
' Authors : Mathieu Jacomy
' Website : http://www.gephi.org
' 
' This file is part of Gephi.
' 
' DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
' 
' Copyright 2011 Gephi Consortium. All rights reserved.
' 
' The contents of this file are subject to the terms of either the GNU
' General Public License Version 3 only ("GPL") or the Common
' Development and Distribution License("CDDL") (collectively, the
' "License"). You may not use this file except in compliance with the
' License. You can obtain a copy of the License at
' http://gephi.org/about/legal/license-notice/
' or /cddl-1.0.txt and /gpl-3.0.txt. See the License for the
' specific language governing permissions and limitations under the
' License.  When distributing the software, include this License Header
' Notice in each file and include the License files at
' /cddl-1.0.txt and /gpl-3.0.txt. If applicable, add the following below the
' License Header, with the fields enclosed by brackets [] replaced by
' your own identifying information:
' "Portions Copyrighted [year] [name of copyright owner]"
' 
' If you wish your version of this file to be governed by only the CDDL
' or only the GPL Version 3, indicate your decision by adding
' "[Contributor] elects to include this software in this distribution
' under the [CDDL or GPL Version 3] license." If you do not indicate a
' single choice of license, a recipient has the option to distribute
' your version of this file under either the CDDL, the GPL Version 3 or
' to extend the choice of license to its licensees as provided above.
' However, if you add GPL Version 3 code and therefore, elected the GPL
' Version 3 license, then the option applies only if the new code is
' made subject to such option by the copyright holder.
' 
' Contributor(s):
' 
' Portions Copyrighted 2011 Gephi Consortium.
' 

Namespace layout

    ''' <summary>
    ''' @author Mathieu Jacomy
    ''' </summary>
    Public Class LabelAdjust

        'Settings
        Private RadiusScale As Single = 1.1F
        'Graph size
        Private XMin As Single
        Private XMax As Single
        Private YMin As Single
        Private YMax As Single

        Dim Converged As Boolean

        Public MaxIterations As Integer = 1000
        Public Canvas As SizeF

        Public Sub ResetPropertiesValues()
            Speed = 1
            RadiusScale = 1.1F
            AdjustBySize = True
        End Sub

        Public Sub Solve(nodes As Node(), labels As Dictionary(Of Node, TextProperties))
            Dim layoutData As LabelAdjustLayoutData

            'Reset Layout Data
            For Each n As Node In nodes
                If n.LayoutData Is Nothing OrElse Not (TypeOf n.LayoutData Is LabelAdjustLayoutData) Then
                    n.LayoutData = New LabelAdjustLayoutData()
                End If
                layoutData = n.LayoutData
                layoutData.Freeze = 0
                layoutData.Dx = 0
                layoutData.Dy = 0
            Next

            Converged = False

            Dim i As Integer = 0

            Do While Not Converged
                Call Iterate(nodes, labels)

                If i > MaxIterations Then
                    Exit Do
                Else
                    i += 1
                End If
            Loop
        End Sub

        ''' <summary>
        ''' one iteration
        ''' </summary>
        ''' <param name="nodes"></param>
        ''' <param name="labels"></param>
        Private Sub Iterate(nodes As Node(), labels As Dictionary(Of Node, TextProperties))
            ' Get xmin, xmax, ymin, ymax
            XMin = Single.MaxValue
            XMax = Single.Epsilon
            YMin = Single.MaxValue
            YMax = Single.Epsilon

            Dim correctNodes As New List(Of Node)()
            For Each n As Node In nodes
                Dim x As Single = n.X()
                Dim y As Single = n.Y()
                Dim t As TextProperties = labels(n)
                Dim w As Single = t.Width
                Dim h As Single = t.Height
                Dim radius As Single = n.Size / 2.0F

                If w > 0 AndAlso h > 0 Then
                    ' Get the rectangle occupied by the node (size + label)
                    Dim nxmin = std.Min(x - w / 2, x - radius)
                    Dim nxmax = std.Max(x + w / 2, x + radius)
                    Dim nymin = std.Min(y - h / 2, y - radius)
                    Dim nymax = std.Max(y + h / 2, y + radius)

                    ' Update global boundaries
                    XMin = std.Min(XMin, nxmin)
                    XMax = std.Max(XMax, nxmax)
                    YMin = std.Min(YMin, nymin)
                    YMax = std.Max(YMax, nymax)

                    correctNodes.Add(n)
                End If
            Next

            If correctNodes.Count = 0 OrElse XMin = XMax OrElse YMin = YMax Then
                Converged = True
                Return
            End If

            Dim timeStamp As Long = 1
            Dim someCollision = False

            'Add all nodes in the quadtree
            Dim quadTree As New QuadTree(correctNodes.Count, (XMax - XMin) / (YMax - YMin)) With {
                .XMax = Me.XMax,
                .XMin = Me.XMin,
                .YMax = Me.YMax,
                .YMin = Me.YMin
            }

            For Each n As Node In correctNodes
                quadTree.Add(n, labels)
            Next

            'Compute repulsion - with neighbours in the 8 quadnodes around the node
            For Each n As Node In correctNodes
                timeStamp += 1
                Dim layoutData As LabelAdjustLayoutData = n.LayoutData
                Dim quad = quadTree.GetQuadNode(layoutData.LabelAdjustQuadNode)

                'Repulse with adjacent quad - but only one per pair of nodes, timestamp is guaranteeing that
                For Each neighbour As Node In quadTree.GetAdjacentNodes(quad.row, quad.col)
                    Dim neighborLayoutData As LabelAdjustLayoutData = neighbour.LayoutData
                    If neighbour IsNot n AndAlso neighborLayoutData.Freeze < timeStamp Then
                        Dim collision = Me.Repulse(n, neighbour, labels)
                        someCollision = someCollision OrElse collision
                    End If
                    neighborLayoutData.Freeze = timeStamp 'Use the existing freeze float variable to set timestamp
                Next
            Next

            If Not someCollision Then
                Converged = True
            Else
                ' apply forces
                For Each n As Node In correctNodes
                    Dim layoutData As LabelAdjustLayoutData = n.LayoutData
                    If Not n.Fixed Then
                        layoutData.Dx *= Speed
                        layoutData.Dy *= Speed
                        Dim x As Single = n.X() + layoutData.Dx
                        Dim y As Single = n.Y() + layoutData.Dy

                        If x < 0 Then x = 0
                        If y < 0 Then y = 0
                        If x + labels(n).Width > Canvas.Width Then
                            x = Canvas.Width - labels(n).Width
                        End If
                        If y + labels(n).Height > Canvas.Height Then
                            y = Canvas.Height - labels(n).Height
                        End If

                        n.X = x
                        n.Y = y
                    End If
                Next
            End If
        End Sub

        Private Function Repulse(n1 As Node, n2 As Node, labels As Dictionary(Of Node, TextProperties)) As Boolean
            Dim collision = False
            Dim n1x As Single = n1.X()
            Dim n1y As Single = n1.Y()
            Dim n2x As Single = n2.X()
            Dim n2y As Single = n2.Y()
            Dim t1 As TextProperties = labels(n1)
            Dim t2 As TextProperties = labels(n2)
            Dim n1w As Single = t1.Width
            Dim n2w As Single = t2.Width
            Dim n1h As Single = t1.Height
            Dim n2h As Single = t2.Height
            Dim n2Data As LabelAdjustLayoutData = n2.LayoutData

            Dim n1xmin = n1x - 0.5 * n1w
            Dim n2xmin = n2x - 0.5 * n2w
            Dim n1ymin = n1y - 0.5 * n1h
            Dim n2ymin = n2y - 0.5 * n2h
            Dim n1xmax = n1x + 0.5 * n1w
            Dim n2xmax = n2x + 0.5 * n2w
            Dim n1ymax = n1y + 0.5 * n1h
            Dim n2ymax = n2y + 0.5 * n2h

            'Sphere repulsion
            If AdjustBySize Then
                Dim xDist As Double = n2x - n1x
                Dim yDist As Double = n2y - n1y
                Dim dist = std.Sqrt(xDist * xDist + yDist * yDist)
                Dim sphereCollision As Boolean = dist < RadiusScale * (n1.Size + n2.Size)
                If sphereCollision Then
                    Dim f As Double = 0.1 * n1.Size / dist
                    If dist > 0 Then
                        n2Data.Dx = CSng(n2Data.Dx + xDist / dist * f)
                        n2Data.Dy = CSng(n2Data.Dy + yDist / dist * f)
                    Else
                        n2Data.Dx = CSng(n2Data.Dx + 0.01 * (0.5 - randf.NextDouble))
                        n2Data.Dy = CSng(n2Data.Dy + 0.01 * (0.5 - randf.NextDouble))
                    End If
                    collision = True
                End If
            End If

            Dim upDifferential = n1ymax - n2ymin
            Dim downDifferential = n2ymax - n1ymin
            Dim labelCollisionXleft = n2xmax - n1xmin
            Dim labelCollisionXright = n1xmax - n2xmin

            If upDifferential > 0 AndAlso downDifferential > 0 Then ' Potential collision
                If labelCollisionXleft > 0 AndAlso labelCollisionXright > 0 Then ' Collision
                    If upDifferential > downDifferential Then
                        ' N1 pushes N2 up
                        n2Data.Dy = CSng(n2Data.Dy - 0.02 * n1h * (0.8 + 0.4 * randf.NextDouble))
                        collision = True
                    Else
                        ' N1 pushes N2 down
                        n2Data.Dy = CSng(n2Data.Dy + 0.02 * n1h * (0.8 + 0.4 * randf.NextDouble))
                        collision = True
                    End If
                    If labelCollisionXleft > labelCollisionXright Then
                        ' N1 pushes N2 right
                        n2Data.Dx = CSng(n2Data.Dx + 0.01 * (n1h * 2) * (0.8 + 0.4 * randf.NextDouble))
                        collision = True
                    Else
                        ' N1 pushes N2 left
                        n2Data.Dx = CSng(n2Data.Dx - 0.01 * (n1h * 2) * (0.8 + 0.4 * randf.NextDouble))
                        collision = True
                    End If
                End If
            End If

            Return collision
        End Function

        Public Overridable Property Speed As Double = 1

        Public Overridable Property AdjustBySize As Boolean = True

    End Class

End Namespace
